voting-contract/src/lib.rs

```diff
@@ -0,0 +1,375 @@
+use borsh::{BorshDeserialize, BorshSerialize};
+use near_sdk::collections::Map;
+use near_sdk::{env, near_bindgen, AccountId, Balance, BlockHeight, EpochHeight};
+use serde::Serialize;
+use std::collections::{BTreeMap, HashMap};
+
+#[global_allocator]
+static ALLOC: wee_alloc::WeeAlloc = wee_alloc::WeeAlloc::INIT;
+
+const MIN_HEIGHT_HORIZON: u64 = 432000;
+const MAX_HEIGHT_HORIZON: u64 = MIN_HEIGHT_HORIZON * 100;
+
+type ProposalId = u64;
+
+#[derive(BorshDeserialize, BorshSerialize, Eq, PartialEq, Debug, Serialize)]
+pub struct Proposal {
+    /// Proposed height to reset the network.
+    proposed_height: BlockHeight,
+    /// Human-readable description of this proposal.
+    description: String,
+}
+
+#[derive(BorshDeserialize, BorshSerialize)]
+pub struct AccountVotes {
+    votes: HashMap<ProposalId, Balance>,
+    total_voted_balance: Balance,
+    account_stake: Balance,
+}
+
+impl AccountVotes {
+    fn add_vote(&mut self, proposal_id: ProposalId, stake: Balance) {
+        self.total_voted_balance =
+            self.total_voted_balance + stake - *self.votes.get(&proposal_id).unwrap_or(&0);
+        self.votes.insert(proposal_id, stake);
+    }
+
+    fn withdraw_vote(&mut self, proposal_id: ProposalId) {
+        let stake = *self.votes.get(&proposal_id).unwrap_or(&0);
+        if self.total_voted_balance < stake {
+            env::panic(
+                format!(
+                    "trying to withdraw {} but only has {}",
+                    self.total_voted_balance, stake
+                )
+                .as_bytes(),
+            );
+        }
+        self.total_voted_balance -= stake;
+    }
+}
+
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize)]
+pub struct Votes {
```

Comment:
I would suggest to restructure this contract per Proposal.
The contract should have a top-level map of proposals. When a proposal is resolved, you can remove all voting information, because it's now irrelevant from the Proposal. But you still have to keep the proposal.
When proposal expires, you can also remove voting information relevant to the expired proposal.

Reply:
What do you mean by "proposal is resolved"